### ISSUE TITLE
Explore options for workflow orchestration tools

### ISSUE THREAD

**Issue by damonbayer** [CDCgov/pyrenew-hew#48]
E.g. make or airflow

@dylanhmorris @gvegayon @SamuelBrand1 I think you all had something to say about this. Feel free to drop comments here.

**Comment by SamuelBrand1** [comment]
My 2p based on not deep understanding is that I like the look of `nextflow`. From various blogposts and my own efforts with nextflow. 

Its very portable and easy to set up (I've tried it), and easy to run with multiple languages. Its seems like compute switching from local dev to sending to azure is straightforward. Requires JAVA to run the executable though.

airflow seems to have a steeper learning curve, but has a beautiful looking control panel.

Make looks pretty ok for the kind of pipelines we might be doing.

**Comment by SamuelBrand1** [comment]
`airflow` is significantly more popular in industry but is less popular that `nextflow` in bioinformatic pipelines.

**Comment by seabbs** [comment]
The real benefit of airflow and nextflow are that they promise to be backend agnostic and so in the longer term you could create a pipeline that could be run locally/in the cloud without having to create infra to dispatch to batch. 

If there is a hurry here potentially there might not be time for that but its also unclear to me how long it would take. 

I think in terms of future recruiting people then its easier to believe that we might have people with `airflow` experience but as Sam says I think nextflow is easier to use/get set up on.

**Comment by damonbayer** [comment]
Two relevant links I found in old Teams chats:

https://techcommunity.microsoft.com/t5/azure-data-factory-blog/introducing-workflow-orchestration-manager-powered-by-apache/ba-p/3730151

https://www.nextflow.io/docs/latest/azure.html

**Comment by SamuelBrand1** [comment]
Unless we have someone with more knowledge coming onto this issue, I suspect best way forward is to create a trial version of a pipeline using each of:

- `airflow`
- `nextflow`
- `make` (this exists basically)

If this is tricky given our tech capability for one of the above thats useful information.

**Comment by gvegayon** [comment]
I think all those would be good solutions, as long as they are properly implemented. That said, here is what I would look for:

1. Something that **currently** works well in Azure and locally without much hassle.
2. Something that can play well with containers, ideally like what GHA does, i.e., runs on a container.
3. Should be well documented and tested. I prefer robustness over bleeding-edge stuff.
4. Reiterating 1 and 3, a current problem I see with Azure is documentation. There's lots of things that you kind of have to guess, so I would pay attention to documentation integrating the framework with other tools like AWS, GC, Azure, and GHA.

**Comment by damonbayer** [comment]
I have also come across [`snakemake`](https://snakemake.github.io/), which supports azure batch: https://techcommunity.microsoft.com/t5/healthcare-and-life-sciences/introducing-snakemake-for-azure-batch/ba-p/3850649

**Comment by gvegayon** [comment]
> I have also come across [`snakemake`](https://snakemake.github.io/), which supports azure batch: https://techcommunity.microsoft.com/t5/healthcare-and-life-sciences/introducing-snakemake-for-azure-batch/ba-p/3850649

Unsure about that... they have a tests badge failing on green, so I wouldn't trust them xD. Now, in a more serious tone, one thing I forgot to mention is dependencies. Any time I think about working with Python, I get the impression that you will need a bazillion libraries for the thing to work. This has been an issue with `renv`, which I have seen creating some trouble across CFA projects.

**Comment by damonbayer** [comment]
Overall, I think I am most intrigued by airflow. I think there could be a future where all the routine NNH/STF work is orchestrated on airflow and we all have access to a nice website where we can see the status of our jobs.

**Comment by SamuelBrand1** [comment]
Is this issue now a duplicate of #46 ? Because `mlflow` being suggested there does seem to do workflow orchestration.

**Comment by seabbs** [comment]
So I think the idea is the jobs are managed by airflow and mlflow is used within that for job tracking and diagnostics (this is the set of tools I also agree with in the absence of ML or bayes in industry people being available to ask (I think if we can we should do this)

**Comment by SamuelBrand1** [comment]
Right, that also makes sense. The `mlflow` plugin to Azure seems reasonable https://learn.microsoft.com/en-us/azure/machine-learning/how-to-deploy-mlflow-models?view=azureml-api-2&tabs=azureml 

So the mental model could be:

1. `mlflow` is tracking experiments on specific states and maybe some optimisation post-processing (e.g. forecast reconciliation).
2. `airflow`/whatever is automating the batch over states?

**Comment by damonbayer** [comment]
> Is this issue now a duplicate of #46 ? Because `mlflow` being suggested there does seem to do workflow orchestration.

I don't completely understand the separation of responsibilities yet, but if you google "MLflow and airflow" there are tons of resources about using them together.

**Comment by SamuelBrand1** [comment]
> > Is this issue now a duplicate of #46 ? Because `mlflow` being suggested there does seem to do workflow orchestration.
> 
> I don't completely understand the separation of responsibilities yet, but if you google "MLflow and airflow" there are tons of resources about using them together.

Yes, I see that now.

**Comment by damonbayer** [comment]
@dylanhmorris is planning to explore Airflow this week.

**Comment by seabbs** [comment]
I also just found `MetaFlow`: https://metaflow.org/

**Comment by damonbayer** [comment]
> I also just found `MetaFlow`: https://metaflow.org/

I also thought that looked promising and shared with @dylanhmorris on Teams. They have an interactive demo here, which I thought was cool https://outerbounds.com/docs/sandbox/
